### PR TITLE
supervisor tracing support

### PR DESCRIPTION
**Pull request by kraman**
Connect to strongloop-internal/scrum-nodeops#330

### CODE REVIEW

File: bin/sl-run.js [hunk 1]
```diff
@@ -17,6 +17,14 @@ var agentOptions = {
   logger: config.logger,
 };
 
+var tracer = require('concurix');
+var tracerOptions = {
+  accountKey: config.clustered.toString()+process.pid.toString(),
+  archiveInterval: 60000,
+  api_host: 'localhost', // XXX(seto) get it from Hotname parameter in Deploy
```
Comment (sam-github) [inline bin/sl-run.js]: Need to replace HTTP with node IPC. We'll talk more, but see lib/express-records.js for example.

File: bin/sl-run.js [hunk 2]
```diff
@@ -17,6 +17,14 @@ var agentOptions = {
   logger: config.logger,
 };
 
+var tracer = require('concurix');
```
Comment (sam-github) [inline bin/sl-run.js]: should be optional, probably controlled by same --profile/--no-profile flag that controls agent

File: bin/sl-run.js [hunk 3]
```diff
@@ -39,6 +48,7 @@ switch(config.profile) {
     // warning messages. If an app is missing a name, profile may still fail
     // to start, so drop-through to start(). We must re-supply options.
     if (agent().config.key)
+      tracer(tracerOptions);
```
Comment (sam-github) [inline bin/sl-run.js]: syntactially invalid! and in wrong spot, tracer has nothing to  do with our hosted strongops, should move to just before agent().start()

**Comment by sam-github** [comment]
OK for POC. We may be able to merge this to a `v2.x` branch (after a bit of cleanup) so it can be available on staging for @seanbrookes, who doesn't care about niceties like whether its using HTTP or node IPC, as long as he can hit the REST APIs from arc.

**Comment by Setogit** [comment]
@seanbrookes  has got this stuff working locally for his Arc side of the work using REST API.  He is not blocked with this.

**Comment by chandadharap** [comment]
Leave  this open until integration is complete even if Sean is unblocked for his SPIKE.

File: bin/sl-run.js (hunk 2 above)
Comment (Setogit) [inline bin/sl-run.js]: It was made parallel to 'var agent.'  Shall we make just "var trace" conditional on config.profile or make both agent declaration and tracer declaration conditional on config.profile?

Comment (sam-github) [inline bin/sl-run.js]: so monkey patching doesn't occur until tracer is called?

File: bin/sl-run.js [hunk 4]
```diff
@@ -17,6 +17,14 @@ var agentOptions = {
   logger: config.logger,
 };
 
+var tracer = require('concurix');
+var tracerOptions = {
+  accountKey: agent().config.appName,
```
Comment (sam-github) [inline bin/sl-run.js]: agent() shouldn't be called if we aren't profiling, its a "lazy loader" (see `lib/agent.js`). I'd suggest wrapping this up as:
```
function trace() {
   var tracerOptions = {
     // ...
  };
  tracer(traceOptions);
}
```

And calling `trace()` in the switch statement.

**Comment by Setogit** [comment]
Connect to strongloop-internal/scrum-nodeops#344

File: bin/sl-run.js [hunk 5]
```diff
@@ -18,7 +18,14 @@ var agentOptions = {
...
 
-switch (config.profile) {
+var tracer = require('concurix');
+var tracerOptions = {
+  archiveInterval: 20000,
+  accountKey: agent().config.appName,
+  useHttp: false
+};
+
+switch(config.profile) {
```
Comment (sam-github) [inline bin/sl-run.js]: Please run `npm lint`. switch needs a single SPC after it.

File: lib/config.js [hunk 6]
```diff
@@ -361,14 +364,20 @@ config.start = function start() {
 };
 
 config.sendMetrics = sendMetrics;
+
```
Comment (sam-github) [inline lib/config.js]: unrelated whitespace change

File: lib/config.js [hunk 7]
```diff
@@ -361,14 +364,20 @@ config.start = function start() {
...
 config.sendExpressRecords = function(parentCtl) {
   if (!this.clustered) return;
   sendExpressRecords(parentCtl);
 };
+
```
Comment (sam-github) [inline lib/config.js]: unrelated whitespace change

File: lib/trace-object.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    worker.on('message', function forwardToParentIfTraceObject(msg) {
+      if (msg.cmd !== 'trace:object') return;
+      debug('master received', msg.record);
+      parentCtl.notify(msg);
+    });
+  });
+};
+
+function forwardTraceToMaster(record) {
+  debug('received %j', record);
```
Comment (sam-github) [inline lib/trace-object.js]: record can be quite large, right? many K of data, maybe larger? If it's json, you can write this as

```
debug('received', debug.json(record))
```

See lib/debug.js for definition of debug.json.

If its binary, I suggest truncating it if over a 60 or so bytes long.

File: bin/sl-run.js [hunk 9]
```diff
@@ -18,7 +18,14 @@ var agentOptions = {
   logger: config.logger,
 };
 
-switch (config.profile) {
+var tracer = require('concurix');
```
Comment (sam-github) [inline bin/sl-run.js]: Use `var tracer = require('./lib/tracer')`.

File: bin/sl-run.js (hunk 5 above)
Comment (sam-github) [inline bin/sl-run.js]: sorry, `npm run lint`

**Comment by sam-github** [comment]
Looks good, a few small comments.

Don't merge until we are done all of https://github.com/strongloop-internal/scrum-nodeops/issues/344

File: lib/tracer.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+// be loaded from the CWD, and the CWD isn't the directory of the application
+// until run has had a chance to chdir() into it.
+
+module.exports = tracer;
+
+function tracer() {
+  // XXX sl-run.js sets the agent logger, and the quiet option, that won't
+  // happen if strongops isn't being used, unless we figure out a way to do it
+  // here.
+  return require('concurix')();
```
Comment (sam-github) [inline lib/tracer.js]: Ah, I see why you made concurix a singleton. Can you not implement the singleton here?

If there is some subtleties, lets talk.

File: lib/config.js (hunk 6 above)
Comment (Setogit) [inline lib/config.js]: removed the whitespace.

File: lib/config.js (hunk 7 above)
Comment (Setogit) [inline lib/config.js]: removed the whitespace.

File: lib/trace-object.js (hunk 8 above)
Comment (Setogit) [inline lib/trace-object.js]: yes, record can be tens of MB.  lib/debug.js is handy.
done.

File: bin/sl-run.js (hunk 9 above)
Comment (Setogit) [inline bin/sl-run.js]: done.  fixed all lint problems and style errors in strong-supervisor.

File: bin/sl-run.js (hunk 5 above)
Comment (Setogit) [inline bin/sl-run.js]: fixed all the problem and style errors that "npm run list" detected in strong-supervisor.

**Comment by Setogit** [comment]
moved tracer singleton pattern from node-concurix to tracer.js and fixed all lint issues in strong-supervisor.

**Comment by Setogit** [comment]
Connect to strongloop-internal/scrum-nodeops#330

**Comment by sam-github** [comment]
Sorry, @Setogit, this repo was using eslint from github while it waited for a bug fix in eslint to be published. Now that it is, we can use published eslint.

It wasn't my intention that you fix lint warning outside of your own code, there were not supposed to be any, but there was because eslint master is changing.

Anyhow, the master now lints clean, can you undo the changes you made in 7a7b314 that don't relate to your code, and rebase on to master, please? Specifically, we prefer the trailing comma.

File: lib/tracer.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+// be loaded from the CWD, and the CWD isn't the directory of the application
+// until run has had a chance to chdir() into it.
+var assert = require('assert');
+
+module.exports = tracer;
+
+var TRACER = null;
+
+function tracer(options) {
+  // XXX sl-run.js sets the agent logger, and the quiet option, that won't
```
Comment (sam-github) [inline lib/tracer.js]: cut-n-paste comment, it doesn't apply to tracer, which has no logger or quiet option.

**Comment by sam-github** [comment]
Other than the cut-n-paste error, and the unfortunate removals of comma everywhere, this LGTM.

But don't merge until complete feature is done.

**Comment by Setogit** [comment]
lint changes reverted.  review please?

File: lib/config.js (hunk 6 above)
Comment (sam-github) [inline lib/config.js]: I still see it here, is it just me/github?

File: lib/config.js (hunk 7 above)
Comment (sam-github) [inline lib/config.js]: i still see it

File: lib/tracer.js (hunk 11 above)
Comment (sam-github) [inline lib/tracer.js]: You didn't remove the comment?

**Comment by sam-github** [comment]
> can you undo the changes you made in 7a7b314 that don't relate to your code, and rebase on to master, please?

You'll need to rebase onto master and do `npm run lint`

File: bin/sl-run.js (hunk 5 above)
Comment (sam-github) [inline bin/sl-run.js]: didn't add SPC after switch

**Comment by sam-github** [comment]
There are unresolved comments, see https://github.com/strongloop/strong-supervisor/pull/98/files

File: lib/tracer.js (hunk 11 above)
Comment (Setogit) [inline lib/tracer.js]: aha, done. sorry!

File: lib/config.js (hunk 6 above)
Comment (Setogit) [inline lib/config.js]: done.

File: lib/config.js (hunk 7 above)
Comment (Setogit) [inline lib/config.js]: done.

File: bin/sl-run.js (hunk 5 above)
Comment (Setogit) [inline bin/sl-run.js]: I think i did once and reverted it because that's not the code area I touched this time.  in any case, added it.

**Comment by Setogit** [comment]
updated.

**Comment by Setogit** [comment]
Fixed the two 'npm run lint' errors:

    bin/sl-run.js
      28:0  error  Keyword "switch" must be followed by whitespace  space-after-keywords
      29:1  error  Trailing spaces not allowed                      no-trailing-spaces

**Comment by kraman** [comment]
@slnode test please

File: package.json [hunk 12]
```diff
@@ -41,6 +41,7 @@
     "test": "mocha -R spec"
   },
   "dependencies": {
+    "concurix": "^1.2.0",
```
Comment (sam-github) [inline package.json]: should be strong-trace, now

**Comment by sam-github** [comment]
@Setogit LGTM. please squash into a single commit now, but don't merge yet.

**Comment by kraman** [comment]
Squashed and added `--trace` flag to turn on tracing.

**Comment by kraman** [comment]
@slnode test please

**Comment by sam-github** [comment]
LGTM

**Pull request merged**